### ISSUE TITLE
Fp16 support for softmax

### ISSUE THREAD

**Issue by eric-haibin-lin** [apache/incubator-mxnet#14072]
Currently, given fp16 inputs, nd.softmax/sym.softmax perform reduction in fp16, which losses precision. The reduction should be done in fp32 instead. 

https://github.com/apache/incubator-mxnet/blob/32c9ca74839ae4d275bcf9a027ea0a711373be81/src/operator/nn/softmax-inl.h#L164-L202
pytorch reference: 
https://github.com/zdevito/ATen/blob/a6cc4156fe4abc9e31f62f2bba1a2f68c58b77b7/aten/src/ATen/native/cuda/SoftMax.cu#L43-L55

**Comment by eric-haibin-lin** [comment]
fixed in https://github.com/apache/incubator-mxnet/pull/14098
